pkg/gadgets/dns/tracer/tracer.go

```diff
@@ -0,0 +1,226 @@
+	return *(*uint16)(unsafe.Pointer(&b[0]))
+}
+
+func NewTracer() (*Tracer, error) {
+	spec, err := ebpf.LoadCollectionSpecFromReader(bytes.NewReader(ebpfProg))
+	if err != nil {
+		return nil, fmt.Errorf("failed to load asset: %s", err)
+	}
+
+	t := &Tracer{
+		spec:        spec,
+		attachments: make(map[string]*link),
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Attach(key string, pid uint32, f func(name, pktType string)) error {
+	if _, ok := t.attachments[key]; ok {
+		return fmt.Errorf("key already attached: %q", key)
+	}
+
+	coll, err := ebpf.NewCollectionWithOptions(t.spec, ebpf.CollectionOptions{Programs: ebpf.ProgramOptions{LogSize: ebpf.DefaultVerifierLogSize * 100}})
+	if err != nil {
+		return fmt.Errorf("failed to create BPF collection: %s", err)
+	}
+
+	rd, err := perf.NewReader(coll.Maps[BPF_MAP_NAME], os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("failed to get a perf reader: %s", err)
+	}
+
+	prog, ok := coll.Programs[BPF_PROG_NAME]
+	if !ok {
+		return fmt.Errorf("Failed to find BPF program %q", BPF_PROG_NAME)
+	}
+
+	sockFd, err := openRawSock(pid)
+	if err != nil {
+		return fmt.Errorf("Failed to open raw socket: %s", err)
+	}
+
+	if err := syscall.SetsockoptInt(sockFd, syscall.SOL_SOCKET, SO_ATTACH_BPF, prog.FD()); err != nil {
+		return fmt.Errorf("Failed to attach BPF program: %s", err)
+	}
+
+	l := &link{
+		collection: coll,
+		sockFd:     sockFd,
+		perfRd:     rd,
+		stop:       make(chan struct{}),
+	}
+	t.attachments[key] = l
+
+	go t.listen(rd, f)
+
+	return nil
+}
+
+func parseDNSEvent(rawSample []byte) (ret string, pktType string) {
+	// Parse the perf event entry into an Event structure.
+	name := make([]byte, C.MAX_DNS_NAME)
+	copy(name, rawSample)
+	var dummy C.struct_event_t
+	if len(rawSample) >= int(unsafe.Offsetof(dummy.pkt_type))+1 {
+		pktType = fmt.Sprintf("%d", rawSample[unsafe.Offsetof(dummy.pkt_type)])
+	}
+
+	// Convert name into a string with dots
+	for i := 0; i < C.MAX_DNS_NAME; i++ {
+		length := int(name[i])
+		if length == 0 {
+			break
+		}
+		if i+1+length < C.MAX_DNS_NAME {
+			ret += string(name[i+1:i+1+length]) + "."
+		}
+		i += length
+	}
+	return
+}
+
+func (t *Tracer) listen(rd *perf.Reader, f func(name, pktType string)) {
+	for {
+		record, err := rd.Read()
+		if err != nil {
+			if perf.IsClosed(err) {
+				return
+			}
+			log.Errorf("Error while reading from perf event reader: %s", err)
+		}
+
+		if record.LostSamples != 0 {
+			log.Errorf("Error: perf event ring buffer full, dropped %d samples", record.LostSamples)
+			continue
+		}
+
+		f(parseDNSEvent(record.RawSample))
+	}
+
+}
+
+func (t *Tracer) Dettach(key string) error {
+	if l, ok := t.attachments[key]; ok {
+		close(l.stop)
+		l.perfRd.Close()
+		unix.Close(l.sockFd)
+		l.collection.Close()
+		delete(t.attachments, key)
+		return nil
+	} else {
+		return fmt.Errorf("key not attached: %q", key)
+	}
+}
+
+func (t *Tracer) Close() {
+	for _, l := range t.attachments {
+		l.collection.Close()
```

Comment:
Just it?, don't we have to close the perf ring buffer and the sockfd as done in `Dettach`?

Reply:
Updated.